### ISSUE TITLE
Allow the Databaseds to listen on all network interfaces

### ISSUE THREAD

**Issue by nleclercq** [mliszcz/tango-cs-docker#11]
So far, the tango-cs container is setup so that the database only listens  for connections on the loopback. This is controlled by the _-ORBendPoint giop:tcp::$ORB_PORT_ command line argument in _tango-cs-docker/resources/supervisord.conf_. It would be interesting to be able to bind the process to all (or a specific) interface(s) using a dedicated env. var. (such as ORB_PORT) - e.g. BIND_ADDR. 

On some OS, such as macOS, we can't reach the database unless it listens for connections on all network interfaces - i.e. BIND_ADDR=0.0.0.0 (or at least, the one representing the docker bridge). This due to the fact that, on macOS, the Docker implementation uses a VM to embed and run the containers. 

To summarize, it would be great to have something like:
_-ORBendPoint giop:tcp:$BIND_ADDR:$ORB_PORT_ with BIND_ADDR defaulting to an empty string if not specified.

**Comment by mliszcz** [comment]
Hi @nleclercq. Could you give an example that is not working for you?

I just started tango-cs container connected to three separate bridge networks (one network for connection to the sql database and two networks for client GIOP connections). I can access the server on different interfaces. Also netstat output confirms that DataBaseds listens on all network interfaces.

```bash
$ docker network create tango-back
$ docker network create tango-front-1
$ docker network create tango-front-2

# sql database
$ docker run -it --rm --name mysql --network tango-back -e MYSQL_ROOT_PASSWORD=root -e MYSQL_INITDB_SKIP_TZINFO=1 tangocs/mysql:9.2.2 --sql-mode="" --innodb=OFF --default-storage-engine=MyISAM 

# databaseds
$ docker run -it --rm --name databaseds --network tango-back -e TANGO_HOST=127.0.0.1:10000 -e MYSQL_HOST=mysql:3306 -e MYSQL_USER=tango -e MYSQL_PASSWORD=tango -e MYSQL_DATABASE=tango --entrypoint /usr/local/bin/DataBaseds tangocs/tango-cs:9.3.3-rc1 2 -ORBendPoint giop:tcp::10000

$ docker network connect tango-front-1 databaseds
$ docker network connect tango-front-2 databaseds

# works with if #1
$ docker run -it --rm --network tango-front-1 -e TANGO_HOST=databaseds:10000 --entrypoint /usr/local/bin/tango_admin tangocs/tango-cs:9.3.3-rc1 --server-list
DataBaseds TangoAccessControl TangoTest 

# works with if #2
$ docker run -it --rm --network tango-front-2 -e TANGO_HOST=databaseds:10000 --entrypoint /usr/local/bin/tango_admin tangocs/tango-cs:9.3.3-rc1 --server-list
DataBaseds TangoAccessControl TangoTest 

$ docker exec -i -u 0 databaseds /bin/bash -c 'apt update && apt install -y iproute2 && ip addr'                                                              
# ...
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
30: eth0@if31: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 02:42:ac:14:00:03 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 172.20.0.3/16 brd 172.20.255.255 scope global eth0
       valid_lft forever preferred_lft forever
32: eth1@if33: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 02:42:ac:15:00:02 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 172.21.0.2/16 brd 172.21.255.255 scope global eth1
       valid_lft forever preferred_lft forever
34: eth2@if35: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 02:42:ac:16:00:02 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 172.22.0.2/16 brd 172.22.255.255 scope global eth2
       valid_lft forever preferred_lft forever

$ docker exec -i -u 0 databaseds /bin/bash -c 'apt update && apt install net-tools && netstat -tulpn'
# ...
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 127.0.0.11:40199        0.0.0.0:*               LISTEN      -                   
tcp        0      0 0.0.0.0:42041           0.0.0.0:*               LISTEN      -                   
tcp6       0      0 :::10000                :::*                    LISTEN      -                   
udp        0      0 127.0.0.11:43025        0.0.0.0:*                           -   
```

This is documented here: http://omniorb.sourceforge.net/omni41/omniORB/omniORB008.html

> TCP endpoints have the format:
>    giop:tcp:host:port 
> The host must be a valid host name or IP address for the server machine. It determines the network interface on which the server listens. The port selects the TCP port to listen on, which must be unoccupied. Either the host or port, or both can be left empty. **If the host is empty**, the ORB publishes the IP address of the first non-loopback network interface it can find (or the loopback if that is the only interface), but **listens on all network interfaces**. If the port is empty, the operating system chooses a port.

**Comment by nleclercq** [comment]
Hi @mliszcz, 
I totally missed the last point in the omniORB documentation. Sorry. 
Well, I now have no idea why I can't reach the Datatbase from macOS. I tried many things  - including explicit port forwarding  (i.e. -p 10000:10000) - but nothing works. Do you see any track to follow in order to diagnose the problem?

**Comment by mliszcz** [comment]
> Well, I now have no idea why I can't reach the Datatbase from macOS

Are you trying to access the exposed port from your host os (localhost)? If yes, this will not work because the docker daemon runs within a VM on MacOS. You should access it from that machine.
* If you are using docker-machine, get its IP address from [`docker-machine inspect`](https://docs.docker.com/machine/reference/inspect/) and then try accessing IP:10000
* If you are using docker for mac, ~~a special DNS name should be available (either `docker.for.mac.localhost:10000` or maybe `docker.local:10000`)~~
  Update: DNS name I mentioned above is for accessing host machine from within container. Port forwarding should just work according to https://docs.docker.com/docker-for-mac/networking/.

**Comment by nleclercq** [comment]
My bad! I added the -p (i.e. port forwarding) option at the end of the _docker run_ command. Moving it before the container name fixes the problem. The databaseds is now properly bound to all interface and I can reach it from the host. Thanks for your help. We can close this issue.

docker ps
d70a89d45f88        tangocs/tango-cs:latest   "/bin/sh -c '/usr/lo…"   4 minutes ago       Up 4 minutes        **0.0.0.0:10000->10000/tcp**   tango_container
